File: packages/ksuid-python/langwatch_ksuid/platform.py
"""Platform detection and crypto operations."""

import secrets

from .types import CryptoProvider, PlatformInfo


def detect_platform() -> PlatformInfo:
    """
    Detects the current platform/runtime environment.

    Returns:
        Platform information including flags for browser, Node.js, Bun, and Deno

    Example:
        >>> platform_info = detect_platform()
        >>> if platform_info['is_node']:
        ...     print('Running in Node.js')
    """
    # Python doesn't run in browsers, Node.js, Bun, or Deno
    return {
        "is_browser": False,
        "is_node": False,
        "is_bun": False,
        "is_deno": False,
    }


class PythonCryptoProvider:
    """Crypto provider for Python using secrets module."""

    def get_random_values(self, array: bytearray) -> bytearray:
        """
        Fills the provided array with cryptographically secure random values.

        Args:
            array: The array to fill with random bytes

        Returns:
            The filled array
        """
        array[:] = secrets.token_bytes(len(array))
        return array

    def random_bytes(self, size: int) -> bytes:
        """
        Generates cryptographically secure random bytes.

        Args:
            size: The number of random bytes to generate

        Returns:
            A bytes object filled with random bytes
        """
        return secrets.token_bytes(size)


def get_crypto_provider() -> CryptoProvider:
    """
    Gets the appropriate crypto provider for the current platform.

    Returns:
        A crypto provider with get_random_values method

    Example:
        >>> crypto = get_crypto_provider()
        >>> random_bytes = crypto.get_random_values(bytearray(16))
    """
    return PythonCryptoProvider()


def get_random_bytes(size: int) -> bytes:
    """
    Generates cryptographically secure random bytes.

    Args:
        size: The number of random bytes to generate

    Returns:
        A bytes object filled with random bytes

    Example:
        >>> random_bytes = get_random_bytes(16)
        >>> print(len(random_bytes))  # 16
    """
    crypto = get_crypto_provider()
    return crypto.random_bytes(size)

File: packages/ksuid/src/platform.ts
import type { PlatformInfo, CryptoProvider } from "./types.ts";

/**
 * Detects the current platform/runtime environment.
 * @returns Platform information including flags for browser, Node.js, Bun, and Deno
 */
export function detectPlatform(): PlatformInfo {
  const isBrowser = typeof window !== "undefined" && typeof window.crypto !== "undefined";
  const isNode =
    typeof process !== "undefined" &&
    typeof process.versions !== "undefined" &&
    typeof process.versions.node !== "undefined";
  const isBun =
    typeof process !== "undefined" &&
    typeof process.versions !== "undefined" &&
    typeof process.versions.bun !== "undefined";
  const isDeno = typeof Deno !== "undefined";

  return {
    isBrowser,
    isNode,
    isBun,
    isDeno,
  };
}

/**
 * Gets the appropriate crypto provider for the current platform.
 * @returns A crypto provider with getRandomValues method
 * @throws {Error} If no crypto provider is available
 */
export function getCryptoProvider(): CryptoProvider {
  const platform = detectPlatform();

  if (platform.isBrowser && typeof window !== "undefined" && window.crypto) {
    return {
      getRandomValues: (array: Uint8Array) => window!.crypto.getRandomValues(array),
    };
  }

  if (platform.isNode) {
    try {
      const crypto = require("crypto") as {
        // eslint-disable-next-line no-unused-vars
        randomBytes: (size: number) => Uint8Array;
      };
      return {
        getRandomValues: (array: Uint8Array) => {
          const randomBytes = crypto.randomBytes(array.length);
          array.set(randomBytes);
          return array;
        },
        randomBytes: (size: number) => crypto.randomBytes(size),
      };
    } catch {
      // Fallback to browser crypto if available
      if (typeof crypto !== "undefined" && crypto) {
        return {
          getRandomValues: (array: Uint8Array) => crypto!.getRandomValues(array),
        };
      }
    }
  }

  // For Bun, Deno, and other platforms
  if (typeof crypto !== "undefined" && crypto) {
    return {
      getRandomValues: (array: Uint8Array) => crypto!.getRandomValues(array),
    };
  }

  throw new Error("No crypto provider available");
}

/**
 * Generates cryptographically secure random bytes.
 * @param size - The number of random bytes to generate
 * @returns A Uint8Array filled with random bytes
 */
export function getRandomBytes(size: number): Uint8Array {
  const crypto = getCryptoProvider();
  const array = new Uint8Array(size);
  crypto.getRandomValues(array);
  return array;
}
